airflow/ti_deps/ti_deps.py

```diff
@@ -0,0 +1,364 @@
+        cur_date = datetime.now()
+        next_task_retry_date = ti.end_date + ti.task.retry_delay
+        if not ti.ready_for_retry():
+            yield self.failing_status(
+                reason="Task is not ready for retry yet but will be retried "
+                        "automatically. Current date is {0} and task will be retried "
+                        "at {1}.".format(cur_date.isoformat(),
+                                        next_task_retry_date.isoformat()))
+
+
+class NotSkippedDep(BaseTIDep):
+    def get_dep_statuses(self, ti, session, exec_context):
+        if ti.state == State.SKIPPED:
+            yield self.failing_status(reason="The task instance has been skipped.")
+
+
+class PastDagrunDep(BaseTIDep):
+    """
+    Is the past dagrun in a state that allows this task instance to run, e.g. did
+    this task instance's task in the previous dagrun complete if we are depending on past
+    """
+
+    def get_dep_statuses(self, ti, session, exec_context):
+        # TODODAN probably need logging for the returns or first yielding a success status
+        if exec_context.ignore_depends_on_past:
+            return
+
+        if not ti.task.depends_on_past:
+            return
+
+        # Don't depend on the previous task instance if we are the first task
+        if ti.execution_date == ti.task.start_date:
+            return
+
+        previous_ti = ti.previous_ti
+        if not previous_ti:
+            yield self.failing_status(
+                reason="depends_on_past is true for this task, but the previous task "
+                        "instance has not run yet.")
+
+        if previous_ti.state not in [State.SUCCESS, State.SKIPPED]:
+            yield self.failing_status(
+                reason="depends_on_past is true for this task, but the previous task "
+                        "instance is in the state '{0}' which is not a successful "
+                        "state.".format(previous_ti.state))
+
+        previous_ti.task = ti.task
+        if (ti.task.wait_for_downstream and
+                not previous_ti.are_dependents_done(session=session)):
+            yield self.failing_status(
+                reason="The tasks downstream of the previous task instance haven't "
+                        "completed.")
+
+
+class PoolHasSpaceDep(BaseTIDep):
+    def get_dep_statuses(self, ti, session, exec_context):
+        if ti.pool_full():
+            yield self.failing_status(reason="Task's pool '{0}' is full.".format(ti.pool))
+
+
+class TriggerRuleDep(BaseTIDep):
+    """
+    Determines if a task's upstream tasks are in a state that allows a given task instance
+    to run.
+    """
+
+    def get_dep_statuses(self, ti, session, exec_context):
+        TI = airflow.models.TaskInstance
+        TR = airflow.models.TriggerRule
+
+        # Checking that all upstream dependencies have succeeded
+        if not ti.task.upstream_list or ti.task.trigger_rule == TR.DUMMY:
+            return
+
+        qry = (
```

Comment:
Note: This is exceptionally slow: can take between 0.05-0.2s per task. So a Dag with a lot of tasks will get slow (scheduler loops > 40s). What's worse with the size of the task_instance table this will be incrementally slow as it requires a table scan every time.

I am noting this here as it might have consequences for the whole approach if we would change this to for example the task reporting to the DagRun.